cranelift/codegen/src/isa/riscv64/lower.isle

```diff
@@ -927,17 +918,27 @@
 
 
 ;;;;;  Rules for `select_spectre_guard`;;;;;;;;;
-(rule
-  (lower (has_type r_ty (select_spectre_guard (icmp cc ca @ (value_type cty) cb) a b)))
-  (let
-    ((dst VecWritableReg (alloc_vec_writable r_ty))
-      (r Reg (lower_icmp cc ca cb cty))
-      (_ Unit (emit (MInst.SelectIf $true (vec_writable_clone dst) r a b))))
-    (vec_writable_to_regs dst)))
 
-(rule -1
-  (lower (has_type ty (select_spectre_guard c @ (value_type cty) x y)))
-  (gen_select ty (truthy_to_reg cty (normalize_cmp_value cty c (ExtendOp.Zero))) x y))
+;; SelectSpectreGuard is equivalent to Select, but we should not use a branch based
+;; lowering for it. Instead we use a conditional move based lowering.
+;;
+;; We don't have cmov's in RISC-V either, but we can emulate those using bitwise
+;; operations, which is what we do below.
+(rule (lower (has_type ty (select_spectre_guard cmp @ (value_type cmp_ty) x @ (value_type arg_ty) y)))
+  (let (;; Build a mask that is 0 or -1 depending on the input comparision value.
+        ;; `lower_bmask` handles normalizing the input.
+        (mask Reg (lower_bmask $I64 cmp_ty cmp))
+        ;; We computed the mask were computing the mask as 64bits. But for i128's we need to do
+        ;; the next part on both halves.
+        ;; This still lowers to a normal single reg based lowering if we are lowering < i128
+        (wide_mask ValueRegs (value_regs mask mask))
+        ;; Using the mask above we can select either `x` or `y` by
+        ;; performing a bitwise `and` on both sides and then merging them
+        ;; together. We know that only the bits of one of the sides will be selected.
+        ;; TODO: We can use `andn` here if we have `Zbb`
+        (lhs ValueRegs (gen_and arg_ty x wide_mask))
+        (rhs ValueRegs (gen_and arg_ty y (gen_bnot arg_ty wide_mask))))
```

Comment:
Since `lower_bmask` can already produce an `I128` result, I think the following should be equivalent:
```suggestion
        (wide_mask Reg (lower_bmask ty cmp_ty cmp))
        ;; Using the mask above we can select either `x` or `y` by
        ;; performing a bitwise `and` on both sides and then merging them
        ;; together. We know that only the bits of one of the sides will be selected.
        ;; TODO: We can use `andn` here if we have `Zbb`
        (lhs ValueRegs (gen_and arg_ty x wide_mask))
        (rhs ValueRegs (gen_and arg_ty y (gen_bnot arg_ty wide_mask))))
```